content/en/docs/concepts/configuration/secret.md

```diff
@@ -886,15 +887,29 @@ In this case, `0` means you have created an empty Secret.
 ### Service account token Secrets
 
 A `kubernetes.io/service-account-token` type of Secret is used to store a
-token that identifies a
+token credential that identifies a
 {{< glossary_tooltip text="service account" term_id="service-account" >}}.
+
+Since 1.22, this type of Secret is no longer used to mount credentials into Pods,
+and obtaining tokens via the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
+API is recommended instead of using service account token Secret objects.
+Tokens obtained from the `TokenRequest` API are more secure than ones stored in Secret objects,
+because they have a bounded lifetime and are not readable by other API clients.
+`kubectl create token` can be used to obtain a token from the `TokenRequest` API.
+
+If you are unable to use the `TokenRequest` API to obtain a token,
+and the security exposure of persisting a non-expiring token credential
+in a readable API object is acceptable for your use case, 
+you can create a service account token Secret object.
+
 When using this Secret type, you need to ensure that the
 `kubernetes.io/service-account.name` annotation is set to an existing
-service account name. A Kubernetes
-{{< glossary_tooltip text="controller" term_id="controller" >}} fills in some
-other fields such as the `kubernetes.io/service-account.uid` annotation, and the
-`token` key in the `data` field, which is set to contain an authentication
-token.
+service account name. If you are creating both the ServiceAccount and
+the Secret objects, that means you should create the ServiceAccount first.
```

Comment:
Does it error if you don’t? Will it work eventually if they apply in the other order?

Reply:
If you create the secret first it's a race with the service account creation whether the token controller will observe it, check that there's no corresponding service account, and delete it as an orphaned secret.

Reply:
This is gonna be problematic when generating them with GitOps. 

I don’t think kubectl even applies orders in a specific order when you apply a directory. 

It’s a good thing I’ve put dependency orchestration for kpt & Config Sync, but I hate to see it required for vanilla resources. It’s terrible UX for declarative assets. 

Is there any way we can add another annotation to make the user intent more clear to avoid the race?

Reply:
> I don’t think kubectl even applies objects in a specific order when you apply a directory.

order is definitely deterministic, and is covered by unit tests to avoid breaking that order: https://github.com/kubernetes/kubernetes/pull/103407

> Is there any way we can add another annotation to make the user intent more clear to avoid the race?

I think requiring the service account to exist is achievable and reasonable. This is similar to setting ownerReferences, where the owning object also has to exist first.

Reply:
> order is definitely deterministic

By file name, yes, but not resource type or anything more fancy. That's why we added apply ordering by resource type (as well as implicit dependency detection for CRD, Namespaces, and explicit dependencies by annotation) in cli-utils.

> I think requiring the service account to exist is achievable and reasonable. This is similar to setting ownerReferences, where the owning object also has to exist first.

Yes. There are several unfortunate examples of precedence, and all of them have to be accounted for when trying to programmatically handle waiting for reconciliation and apply ordering.

We have a number of exceptions hard coded in the kstatus library: 
https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/status/core.go#L22

And there's apply ordering hard coded as well (tho we're planning to replace this with more implicit dependency detection):
https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/ordering/sort.go#L74

It looks like we haven't been handling ServiceAccount or Secret with any exceptions yet. So they'll need to be added. I'm not sure how to detect if a ServiceAccount is reconciled tho. We weren't previously waiting for the Secret or token to be generated, but that probably means it was considered reconciled immediately, since there's no status.

I'm beginning to think all resources should have status and conditions.

Reply:
Straight creation in order is sufficient, the token controller will do a live lookup to verify the service account is absent before cleaning up the token, so you don't have to wait after creating the service account to proceed to the secret.

Reply:
Noted. Adding it to the resource apply order should be sufficient then until we move to async/parallel apply. Then we'll have to add it as an implicit dependency.